In [56]:
# Importing packages
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import re
from functools import reduce
from collections import defaultdict
from sklearn.impute import KNNImputer
from imblearn.over_sampling import SMOTE
from sklearn.pipeline import Pipeline
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer
from sklearn.preprocessing import OrdinalEncoder, OneHotEncoder, MinMaxScaler, StandardScaler
from sklearn.decomposition import TruncatedSVD
from sklearn.cross_decomposition import PLSSVD
from sklearn.compose import ColumnTransformer

# displaying all the rows and columns
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# setting random seed:
RANDOM_SEED = 42

## 1. EDA

In [57]:
# Load the data:
data = pd.read_csv('/Users/bangxixiao/Desktop/python_projects/DATA1030_MIDTERM_PROJECT/data/application_data.csv', 
                   encoding='UTF-8')

In [3]:
print('The credit default data has {} rows and {} columns.'.format(data.shape[0], data.shape[1]))

The credit default data has 307511 rows and 122 columns.


In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 307511 entries, 0 to 307510
Columns: 122 entries, SK_ID_CURR to AMT_REQ_CREDIT_BUREAU_YEAR
dtypes: float64(65), int64(41), object(16)
memory usage: 286.2+ MB


The data has 307511 entries and 122 columns, where 65 columns are identified as float, 41 columns as integer and 16 columns as object (string).

In [5]:
data.head()

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0      100002       1         Cash loans           M            N   
1      100003       0         Cash loans           F            N   
2      100004       0    Revolving loans           M            Y   
3      100006       0         Cash loans           F            N   
4      100007       0         Cash loans           M            N   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               Y             0          202500.0    406597.5      24700.5   
1               N             0          270000.0   1293502.5      35698.5   
2               Y             0           67500.0    135000.0       6750.0   
3               Y             0          135000.0    312682.5      29686.5   
4               Y             0          121500.0    513000.0      21865.5   

   AMT_GOODS_PRICE NAME_TYPE_SUITE NAME_INCOME_TYPE  \
0         351000.0   Unaccompanied          Working   
1        1129500.0          Family    State servant   
2         135000.0   Unaccompanied          Working   
3         297000.0   Unaccompanied          Working   
4         513000.0   Unaccompanied          Working   

             NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  \
0  Secondary / secondary special  Single / not married  House / apartment   
1               Higher education               Married  House / apartment   
2  Secondary / secondary special  Single / not married  House / apartment   
3  Secondary / secondary special        Civil marriage  House / apartment   
4  Secondary / secondary special  Single / not married  House / apartment   

   REGION_POPULATION_RELATIVE  DAYS_BIRTH  DAYS_EMPLOYED  DAYS_REGISTRATION  \
0                    0.018801       -9461           -637            -3648.0   
1                    0.003541      -16765          -1188            -1186.0   
2                    0.010032      -19046           -225            -4260.0   
3                    0.008019      -19005          -3039            -9833.0   
4                    0.028663      -19932          -3038            -4311.0   

   DAYS_ID_PUBLISH  OWN_CAR_AGE  FLAG_MOBIL  FLAG_EMP_PHONE  FLAG_WORK_PHONE  \
0            -2120          NaN           1               1                0   
1             -291          NaN           1               1                0   
2            -2531         26.0           1               1                1   
3            -2437          NaN           1               1                0   
4            -3458          NaN           1               1                0   

   FLAG_CONT_MOBILE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  CNT_FAM_MEMBERS  \
0                 1           1           0        Laborers              1.0   
1                 1           1           0      Core staff              2.0   
2                 1           1           0        Laborers              1.0   
3                 1           0           0        Laborers              2.0   
4                 1           0           0      Core staff              1.0   

   REGION_RATING_CLIENT  REGION_RATING_CLIENT_W_CITY  \
0                     2                            2   
1                     1                            1   
2                     2                            2   
3                     2                            2   
4                     2                            2   

  WEEKDAY_APPR_PROCESS_START  HOUR_APPR_PROCESS_START  \
0                  WEDNESDAY                       10   
1                     MONDAY                       11   
2                     MONDAY                        9   
3                  WEDNESDAY                       17   
4                   THURSDAY                       11   

   REG_REGION_NOT_LIVE_REGION  REG_REGION_NOT_WORK_REGION  \
0                           0                           0   
1                           0                           0   
2                        

We noticed that there exists a small proportion of missing data.

In [6]:
# describe the data:
data.describe()

SK_ID_CURR         TARGET   CNT_CHILDREN  AMT_INCOME_TOTAL  \
count  307511.000000  307511.000000  307511.000000      3.075110e+05   
mean   278180.518577       0.080729       0.417052      1.687979e+05   
std    102790.175348       0.272419       0.722121      2.371231e+05   
min    100002.000000       0.000000       0.000000      2.565000e+04   
25%    189145.500000       0.000000       0.000000      1.125000e+05   
50%    278202.000000       0.000000       0.000000      1.471500e+05   
75%    367142.500000       0.000000       1.000000      2.025000e+05   
max    456255.000000       1.000000      19.000000      1.170000e+08   

         AMT_CREDIT    AMT_ANNUITY  AMT_GOODS_PRICE  \
count  3.075110e+05  307499.000000     3.072330e+05   
mean   5.990260e+05   27108.573909     5.383962e+05   
std    4.024908e+05   14493.737315     3.694465e+05   
min    4.500000e+04    1615.500000     4.050000e+04   
25%    2.700000e+05   16524.000000     2.385000e+05   
50%    5.135310e+05   24903.000000     4.500000e+05   
75%    8.086500e+05   34596.000000     6.795000e+05   
max    4.050000e+06  258025.500000     4.050000e+06   

       REGION_POPULATION_RELATIVE     DAYS_BIRTH  DAYS_EMPLOYED  \
count               307511.000000  307511.000000  307511.000000   
mean                     0.020868  -16036.995067   63815.045904   
std                      0.013831    4363.988632  141275.766519   
min                      0.000290  -25229.000000  -17912.000000   
25%                      0.010006  -19682.000000   -2760.000000   
50%                      0.018850  -15750.000000   -1213.000000   
75%                      0.028663  -12413.000000    -289.000000   
max                      0.072508   -7489.000000  365243.000000   

       DAYS_REGISTRATION  DAYS_ID_PUBLISH    OWN_CAR_AGE     FLAG_MOBIL  \
count      307511.000000    307511.000000  104582.000000  307511.000000   
mean        -4986.120328     -2994.202373      12.061091       0.999997   
std          3522.886321      1509.450419      11.944812       0.001803   
min        -24672.000000     -7197.000000       0.000000       0.000000   
25%         -7479.500000     -4299.000000       5.000000       1.000000   
50%         -4504.000000     -3254.000000       9.000000       1.000000   
75%         -2010.000000     -1720.000000      15.000000       1.000000   
max             0.000000         0.000000      91.000000       1.000000   

       FLAG_EMP_PHONE  FLAG_WORK_PHONE  FLAG_CONT_MOBILE     FLAG_PHONE  \
count   307511.000000    307511.000000     307511.000000  307511.000000   
mean         0.819889         0.199368          0.998133       0.281066   
std          0.384280         0.399526          0.043164       0.449521   
min          0.000000         0.000000          0.000000       0.000000   
25%          1.000000         0.000000          1.000000       0.000000   
50%          1.000000         0.000000          1.000000       0.000000   
75%          1.000000         0.000000          1.000000       1.000000   
max          1.000000         1.000000          1.000000       1.000000   

          FLAG_EMAIL  CNT_FAM_MEMBERS  REGION_RATING_CLIENT  \
count  307511.000000    307509.000000         307511.000000   
mean        0.056720         2.152665              2.052463   
std         0.231307         0.910682              0.509034   
min         0.000000         1.000000              1.000000   
25%         0.000000         2.000000              2.000000   
50%         0.000000         2.000000              2.000000   
75%         0.000000         3.000000              2.000000   
max         1.000000        20.000000              3.000000   

       REGION_RATING_CLIENT_W_CITY  HOUR_APPR_PROCESS_START  \
count                307511.000000            307511.000000   
mean                      2.031521                12.063419   
std                       0.502737                 3.265832   
min                       1.000000                 0.000000   
25%                       

The columns could be separated into different categories by looking at the prefix of column names:

AMT: data related to specific numbers such as loan annuity, annuity of previous application and so on.

APARTMENTS: apartment-related information

BASEMENTAREA: basement information about building where the clients live

CHANEL: Through which channel we acquired the client on the previous application

CNT: count data such the number of children the client has

CODE: includes gender and why previous application rejected

COMMONAREA: common area information about building where the clients live

DAYS: includes date-related information such as birth, employment

DEF: clients' social surroundings count

ELEVATORS: living condition of clients

EMERGENCYSTATE: living condition of clients

ENTRANCES: living condition of clients

EXT: normalized score from external data source

FLAG: consists of bool values

FLOORSMAX/FLOORSMIN: living condition of clients

HOUR: data regarding the time, for example, the exact hour the client apply for the loan

LANDAREA: living condition of the clients

LIVE: if the client's contact address match with the work address

LIVINGAPARTMENTS/LIVINGAREA: living condition of clients

NAME: mostly strings data regarding the client's personal information

NFLAG: bool-like data

NONLIVINGAPARTMENT/NONLIVINGAREA: living condition of the clients

OBS: number of observations of clients' social surroundings

OCCUPATION: kind of occupation client has

ORGANIZATION: type of organization where client works

OWN: age of client's car

PRODUCT: detailed product combination of the previous application

RATE: fractional number

REG: bool type data, regarding the match of client's permanent address

SELLERPLACE: selling area of seller place of the previous application

SK_: ID-like data

TARGET: target variable

TOTALAREA: living condition of client

WALLSMATERIAL: living condition of client

WEEKDAY: which day in a week the client applies for a loan

YEARS: living condition of client

In [61]:
(data.isnull().sum(axis=0) / data.shape[0]).sort_values()

SK_ID_CURR                      0.000000
HOUR_APPR_PROCESS_START         0.000000
REG_REGION_NOT_WORK_REGION      0.000000
LIVE_REGION_NOT_WORK_REGION     0.000000
REG_CITY_NOT_LIVE_CITY          0.000000
REG_CITY_NOT_WORK_CITY          0.000000
LIVE_CITY_NOT_WORK_CITY         0.000000
ORGANIZATION_TYPE               0.000000
FLAG_DOCUMENT_21                0.000000
FLAG_DOCUMENT_20                0.000000
FLAG_DOCUMENT_19                0.000000
FLAG_DOCUMENT_18                0.000000
FLAG_DOCUMENT_17                0.000000
FLAG_DOCUMENT_16                0.000000
FLAG_DOCUMENT_15                0.000000
FLAG_DOCUMENT_14                0.000000
FLAG_DOCUMENT_13                0.000000
FLAG_DOCUMENT_12                0.000000
FLAG_DOCUMENT_11                0.000000
FLAG_DOCUMENT_10                0.000000
FLAG_DOCUMENT_9                 0.000000
FLAG_DOCUMENT_8                 0.000000
FLAG_DOCUMENT_7                 0.000000
FLAG_DOCUMENT_6                 0.000000
FLAG_DOCUMENT_5 

In [62]:
drop_columns = []
drop_columns += list(data.columns[data.isnull().sum(axis=0) / data.shape[0] > 0.35])

In [63]:
data['EXT_SOURCE_3'].fillna(data['EXT_SOURCE_3'].median(), inplace=True)

In [64]:
cred_bureau_cols = [
    col 
    for col in data.columns 
    if col.startswith("AMT_REQ_CREDIT_BUREAU_")
]
for col in cred_bureau_cols:
    data[col] = data[col].fillna(0.0)
drop_columns.extend(['AMT_REQ_CREDIT_BUREAU_YEAR', 'DAYS_LAST_PHONE_CHANGE'])
drop_columns.extend(['NAME_TYPE_SUITE', 'OCCUPATION_TYPE'])

In [81]:
data['NAME_INCOME_TYPE'].value_counts()

Working        250482
Retired         55016
Unemployed         19
Student            18
Businessman        10
Name: NAME_INCOME_TYPE, dtype: int64

In [65]:
replace_dict = {
    "Commercial associate" : "Working", 
    "State servant"        : "Working", 
    "Maternity leave"      : "Working", 
    "Pensioner"            : "Retired",
    'Businessman': 'Working',
}
data['NAME_INCOME_TYPE'] = data['NAME_INCOME_TYPE'].replace(replace_dict)


replace_dict = {
    "Secondary / secondary special" : "Secondary_Education",
    "Higher education"              : "Higher_Education",
    "Incomplete higher"             : "Secondary_Education",
    "Lower secondary"               : "Lower_Secondary",
    "Academic degree"               : "Higher Education"
}
data['NAME_EDUCATION_TYPE'] = data['NAME_EDUCATION_TYPE'].replace(replace_dict)


replace_dict = {
    "Single / not married" : "Single",
    "Civil marriage"       : "Married",
    "Separated"            : "Married",
    "Widow"                : "Married",
    "Unknown"              : "Married"
}
data['NAME_FAMILY_STATUS'].replace(replace_dict, inplace=True)


replace_dict = {
    "House / apartment"   : "Owned",
    "With parents"        : "Owned",
    "Municipal apartment" : "Freehold",
    "Office apartment"    : "Freehold",
    "Rented apartment"    : "Rented",
    "Co-op"               : "Rented"
}
data['NAME_HOUSING_TYPE'].replace(replace_dict, inplace=True)

In [66]:
flag_doc_cols = ['FLAG_DOCUMENT_{}'.format(x) for x in range(2, 22)]
data['CNT_DOCUMENT'] = data[flag_doc_cols].sum(axis=1)
data.drop(columns=flag_doc_cols, inplace=True)

In [67]:
data['CODE_GENDER'].replace({"M": 1, "F": 0, "XNA": 1}, inplace=True)
data['CODE_GENDER'] = data['CODE_GENDER'].astype('int64')
data['DAYS_BIRTH'] = data['DAYS_BIRTH'].apply(lambda x: int(x * (-1) / 365))
data['DAYS_EMPLOYED'] = data['DAYS_EMPLOYED'].apply(lambda x: int(x * (-1) / 365))

In [68]:
data.drop(columns=[
    'WEEKDAY_APPR_PROCESS_START', 'HOUR_APPR_PROCESS_START', 'SK_ID_CURR'
], inplace=True)
data.drop(columns=drop_columns, inplace=True)

In [69]:
data.dtypes.sort_values()

TARGET                           int64
LIVE_CITY_NOT_WORK_CITY          int64
REG_CITY_NOT_WORK_CITY           int64
REG_CITY_NOT_LIVE_CITY           int64
LIVE_REGION_NOT_WORK_REGION      int64
REG_REGION_NOT_WORK_REGION       int64
REG_REGION_NOT_LIVE_REGION       int64
REGION_RATING_CLIENT_W_CITY      int64
REGION_RATING_CLIENT             int64
FLAG_EMAIL                       int64
FLAG_CONT_MOBILE                 int64
FLAG_WORK_PHONE                  int64
FLAG_EMP_PHONE                   int64
FLAG_MOBIL                       int64
DAYS_ID_PUBLISH                  int64
DAYS_EMPLOYED                    int64
FLAG_PHONE                       int64
CNT_DOCUMENT                     int64
CODE_GENDER                      int64
CNT_CHILDREN                     int64
DAYS_BIRTH                       int64
AMT_REQ_CREDIT_BUREAU_MON      float64
AMT_REQ_CREDIT_BUREAU_WEEK     float64
AMT_REQ_CREDIT_BUREAU_DAY      float64
AMT_REQ_CREDIT_BUREAU_HOUR     float64
DEF_60_CNT_SOCIAL_CIRCLE 

In [79]:
ORDINAL_FEATURES = ['REGION_RATING_CLIENT', 'REGION_RATING_CLIENT_W_CITY']
CONTINUOUS_FEATURES = ['EXT_SOURCE_2',
                         'EXT_SOURCE_3',
                         'AMT_INCOME_TOTAL',
                         'AMT_CREDIT',
                         'AMT_ANNUITY',
                         'AMT_GOODS_PRICE',
                         'AMT_REQ_CREDIT_BUREAU_MON',
                         'AMT_REQ_CREDIT_BUREAU_DAY',
                         'AMT_REQ_CREDIT_BUREAU_WEEK',
                         'AMT_REQ_CREDIT_BUREAU_QRT',
                         'AMT_REQ_CREDIT_BUREAU_HOUR',
                         'OBS_30_CNT_SOCIAL_CIRCLE',
                         'DEF_30_CNT_SOCIAL_CIRCLE',
                         'OBS_60_CNT_SOCIAL_CIRCLE',
                         'DEF_60_CNT_SOCIAL_CIRCLE',
                         'CNT_CHILDREN',
                         'CNT_FAM_MEMBERS',
                         'CNT_DOCUMENT',
                         'DAYS_BIRTH',
                         'DAYS_EMPLOYED',
                         'DAYS_REGISTRATION',
                         'DAYS_ID_PUBLISH',
                         'REGION_POPULATION_RELATIVE']
CATEGORICAL_FEATURES = ['NAME_CONTRACT_TYPE',
                         'NAME_INCOME_TYPE',
                         'NAME_EDUCATION_TYPE',
                         'NAME_FAMILY_STATUS',
                         'NAME_HOUSING_TYPE',
                         'FLAG_OWN_CAR',
                         'FLAG_OWN_REALTY',
                         'FLAG_EMP_PHONE',
                         'FLAG_WORK_PHONE',
                         'FLAG_PHONE',
                         'FLAG_EMAIL',
                         'CODE_GENDER',
                         'FLAG_MOBIL',
                         'FLAG_CONT_MOBILE',
                         'ORGANIZATION_TYPE',
                         'LIVE_REGION_NOT_WORK_REGION',
                         'LIVE_CITY_NOT_WORK_CITY',
                         'REG_REGION_NOT_LIVE_REGION',
                         'REG_REGION_NOT_WORK_REGION',
                         'REG_CITY_NOT_LIVE_CITY',
                         'REG_CITY_NOT_WORK_CITY']

In [80]:
data = data.dropna()
data[CONTINUOUS_FEATURES + CATEGORICAL_FEATURES + ORDINAL_FEATURES + ['TARGET']].to_csv(
    '/Users/bangxixiao/Desktop/python_projects/DATA1030_MIDTERM_PROJECT/data/data_imputed_2.csv', 
    index=False
)